### ISSUE TITLE
Does not work with anonymous user permissions

### ISSUE THREAD

**Issue by rchrd2** [sandstorm-io/meteor-accounts-sandstorm#9]
I noticed that this doesn't work with anonymous users. For example, my app is receiving the following headers, but the library takes no action, because there isn't a sandstorm-user-id (see https://github.com/sandstorm-io/meteor-accounts-sandstorm/blob/master/server.js#L48).

```
{ connection: 'close',
  'accept-encoding': 'gzip',
  'accept-language': 'en-US,en;q=0.5',
  'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:41.0) Gecko/20100101 Firefox/41.0',
  'x-sandstorm-username': 'Anonymous%20User',
  'x-sandstorm-permissions': 'modify',
  'x-sandstorm-base-path': 'http://35bea8c9f81f01acf2b0ee2914c81cbd.local.sandstorm.io:6080',
  host: '35bea8c9f81f01acf2b0ee2914c81cbd.local.sandstorm.io:6080',
  'x-forwarded-proto': 'http',
  'x-sandstorm-session-id': '0',
  accept: '*/*' }
```

I noticed this bug when I shared an edit link to a TextEditor grain, but the user who received the link could not edit the text.

As I think about this, ideally there wouldn't have to be a new User doc for every anonymous user. So maybe it is better for apps that want to support anonymous users to just check the headers manually. 

I'm not really sure how to proceed. If there are no code changes, a note in the README about anonymous users would be great. 

Do you have any ideas?

**Comment by rchrd2** [comment]
I've had some more time to work on this. I guess the question is, should this library support Anonymous Users and if so, how should it do that? There are some Anonymous User packages for Meteor, which might be useable.

In the meantime, I have found a workaround by checking the headers directly in my app. I used the package `gadicohen:headers@=0.0.30` to make the headers accessible within the application code.

Here's an example:

```
  Meteor.methods({
    /**
     * Note it's important to use an old-school function() {}, so `this` gets
     * bound correctly
     */
    hasPermission: function () {
      var h = headers.get(this);
      var p = h['x-sandstorm-permissions'] || "";
      return p.indexOf('modify') !== -1 || p.indexOf('owner') !== -1;
    },
  });
```

https://github.com/rchrd2/TextEditor/blob/add-user-permissions/texteditor.js#L60

**Comment by paulproteus** [comment]
I would personally say (I'm not the maintainer of this particular library) that, if possible, Meteor.user() should return a User object that has a `.sandstorm` that contains an empty permissions list.

I wonder -- maybe it makes sense to, upon the first visit by an anonymous user, create a pseudo-account for the anonymous user, and then use that same user object for all anonymous users. Since anonymous users can't be assigned permissions, this avoids a storage cost of one ID per anonymous user (doubly so since I think we can't even distinguish anonymoususer1 from anonymoususer2 from anonymoususer1-who-just-clicked-reload).

This has the downside that if one such user manages to modify their user data, it'll be modified for all anonymous users.

So I think what I'd suggest instead is that `Meteor.user()` should return a fake user document that's not stored in the database but at least does contain an empty `permissions` object, and that if apps want to have custom behavior for anonymous users, then they can implement that somehow.

_Or_ `meteor-accounts-sandstorm` should provide a different `hasPermission` function that correctly returns `false` for all permissions, even if there is no User document.

**Comment by kentonv** [comment]
Monkey-patching `Meteor.user()` sounds pretty risky to me as it's used all over Meteor internals and in security-sensitive code.

IMO what we want is a user-independent notion of permissions. Permissions apply to a session, not a user. It should be easy to fetch the permissions for the current DDP connection. And it almost is really easy, except that sockjs only lets whitelisted headers through for security reasons (that don't apply to Sandstorm).

**Comment by ndarilek** [comment]
I've been thinking about this lately and have the following implementation:

`Sandstorm.permissions()`: (everywhere) Returns the permissions associated with the current session.
`Sandstorm.hasPermission(permission, default=false)`: (everywhere) Wrapper to determine whether session has the given permission, returning `default` otherwise.

I see this working everywhere on the client, and in methods/publications on the server. On the client I've implemented this as a ReactiveVar and a simple getter function, set alongside the user creation code. I'm stumbling a bit on the server though, because in method calls `this.connection.httpHeaders` appears to be missing the Sandstorm headers, perhaps due to the DDP security whitelist mentioned above.

Does the headers package circumvent this whitelist? If so, I'll incorporate that workaround into the server-side implementation of this API. If it does then I'd like to add it as a package dependency, and thought I'd check if that was acceptable before submitting a PR that did so.

And can anyone think of a better API than this? I don't particularly like the global singleton but I don't like creating anonymous users all that much either.

I'd like to sort this out and submit a PR in the next couple weeks, because it bugs me that Meteor apps can't support anonymous users with permissions and I have a couple use cases in mind where I want that. :)

**Comment by rchrd2** [comment]
I dug into this problem and also tried Meteor Headers. Unfortunately, the Headers provided by Meteor headers are only available during the first DDP connection. If the client disconnects and reconnects and forms a new DDP connection, those headers are no longer available. 

The permissions need to be stored in some sort of session object, and it turns out the built-in user model is the easiest place to store this, because it is global variable that's available in many functions. 

For my solution, I did end up generating an Anonymous User. I store this user ID in local storage, so that only one Anonymous User is created per connected user. I then store the permissions in this user model, and update these permissions everytime they connect to the application.

Here you can see what is happening on the server:
https://github.com/rchrd2/TextEditor/blob/polymer/server/sandstorm_helpers.js

Here is example client code (apologies is it is confusing, because it is using Polymer with Meteor). Just search for "localStorage":
https://github.com/rchrd2/TextEditor/blob/polymer/public/components/x-main.html

This solution has worked well so far. Perhaps something like this can be ported into this library.

**Comment by kentonv** [comment]
This is fixed in 0.2! See the readme for details.
